src/charm.py

```diff
@@ -165,12 +175,85 @@ def external_url(self) -> str:
+            return None
+
+    def _push(self, path: str, contents: Any):
+        """Push file to container, creating subdirs as necessary."""
+        self._nginx_container.push(path, contents, make_dirs=True, encoding="utf-8")
+
+    def _push_alert_rules(self, alerts: Dict[str, Any]) -> List[str]:
+        """Pushes alert rules from a rules file to the nginx container.
+
+        Args:
+            alerts: a dictionary of alert rule files, fetched from
+                either a metrics consumer or a remote write provider.
+        """
+        paths = []
+        for topology_identifier, rules_file in alerts.items():
+            filename = f"juju_{topology_identifier}.rules"
+            path = f"{RULES_DIR}/{filename}"
+
+            rules = yaml.safe_dump(rules_file)
+
+            self._push(path, rules)
+            paths.append(path)
+            logger.debug("Updated alert rules file %s", filename)
+
+        return paths
+
+    def _push_mimirtool(self):
+        """Copy the `mimirtool` binary to the workload container."""
+        with open("mimirtool", "rb") as f:
+            self._nginx_container.push("/usr/bin/mimirtool", source=f, permissions=0o744)
+
+    def _set_alerts(self):
+        """Create alert rule files for all Mimir consumers."""
+
+        def sha256(hashable: Any) -> str:
+            """Use instead of the builtin hash() for repeatable values."""
+            if isinstance(hashable, str):
+                hashable = hashable.encode("utf-8")
+            return hashlib.sha256(hashable).hexdigest()
+
+        # Get mimirtool if this is the first execution
+        if not self._pull(ALERTS_HASH_PATH):
+            self._push_mimirtool()
+
+        remote_write_alerts = self.remote_write_provider.alerts
+        alerts_hash = sha256(str(remote_write_alerts))
+        alert_rules_changed = alerts_hash != self._pull(ALERTS_HASH_PATH)
+
+        if alert_rules_changed:
+            # Update the alert rules files on disk
+            self._nginx_container.remove_path(RULES_DIR, recursive=True)
+            rules_file_paths: List[str] = self._push_alert_rules(remote_write_alerts)
+            self._push(ALERTS_HASH_PATH, alerts_hash)
+            # Push the alert rules to the Mimir cluster (persisted in s3)
+            mimirtool_output = self._nginx_container.pebble.exec(
+                [
+                    "mimirtool",
+                    "rules",
+                    "sync",
+                    *rules_file_paths,
+                    f"--address={self.external_url}",
+                    "--id=anonymous",  # multitenancy is disabled, the default tenant is 'anonymous'
+                ],
+                encoding="utf-8",
+            )
+            if mimirtool_output.stdout:
+                logger.info(f"mimirtool: {mimirtool_output.stdout.read().strip()}")
+            if mimirtool_output.stderr:
+                logger.info(f"mimirtool (err): {mimirtool_output.stderr.read().strip()}")
```

Comment:
If there is an error in the mimirtool execution... shouldn't we log an error?

```suggestion
            if mimirtool_output.stderr:
                logger.error(f"mimirtool (err): {mimirtool_output.stderr.read().strip()}")
```

Is this enough? 🤔

Reply:
Good point! Yes this should be enough :)